Routing: New URL routing for Joomla

Hackwar opened the pull request:
This is an updated PR from #7615

This PR implements a new routing system for Joomla, based on what has been proposed here: http://www.joomlager.de/crowdfunding

This is a combined PR of the former changes from #5446, #5501, #5502. #5503, #5509, #5599 and #5604.

Please test and comment. For further informations, read the other 7 PRs.

Hackwar committed:
Implementing JComponentRouterAdvanced class

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Implementing JComponentRouterRulesMenu class to find right Itemid

Hackwar committed:
Call-time pass-by-reference has been removed

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrulesitemid

Hackwar committed:
Removing JError, using Exception instead

Hackwar committed:
protecting $name and renaming register() to registerView()

Hackwar committed:
Codestyle and minor bug

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Adding removeRule, getRules and renamed $id to $key in register method

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrulesitemid

Hackwar committed:
Implementing legacy router for com_content

Hackwar committed:
Codestylefix

Hackwar committed:
Implementing legacy router for com_contact

zero-24 committed:
Fix CS issues that was found by Travis

https://travis-ci.org/joomla/joomla-cms/jobs/44966357

Hackwar committed:
Merge pull request #14 from zero-24/patch-25

Fix CS issues that was found by Travis

Hackwar committed:
Implementing legacy router for com_newsfeeds

Hackwar committed:
Codestyle

Hackwar committed:
Codestyle

Hackwar committed:
Implementing legacy router for com_users

Hackwar committed:
Changing component router to retrieve, since the com_content router is pretty complex and throws errors in our testing environment

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_content_router_legacy

Hackwar committed:
Merge branch 'com_content_router_legacy' of github.com:Hackwar/joomla-cms into com_content_router_legacy

Hackwar committed:
Making method names consistent

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Implementing JComponentRouterViewconfiguration for configuration of views in JComponentRouterAdvanced

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Codestyle, smaller improvements, unittests for all component router classes except for JComponentRouterAdvanced

Hackwar committed:
Removing ability to have one view with different names and implementing unittests for JComponentRouterAdvanced

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrulesitemid

Hackwar committed:
Fixes to JComponentRouterRulesMenu

Hackwar committed:
Small fixes and unittests for JComponentRouterRulesMenu

Hackwar committed:
Codestyle

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_content_router_legacy

Hackwar committed:
Implementing new view registering method in com_content

Hackwar committed:
Implementing new view registering method in com_contact

Hackwar committed:
Merge branch 'com_contact_router_legacy' of github.com:Hackwar/joomla-cms into com_contact_router_legacy

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_contact_router_legacy

Hackwar committed:
Codestyle

Hackwar committed:
Implementing new view registering method in com_newsfeeds

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_newsfeed_router_legacy

Hackwar committed:
Merge branch 'com_newsfeed_router_legacy' of github.com:Hackwar/joomla-cms into com_newsfeed_router_legacy

Hackwar committed:
Implementing new view registering method in com_users

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_users_router_legacy

Hackwar committed:
Implementing "no Itemid" behavior

Hackwar committed:
Unittest improvements

Hackwar committed:
Implementing standard routing behavior for components

Hackwar committed:
Adding get<View>Slug() and get<View>Key() methods to JComponentRouterAdvanced

Hackwar committed:
Updating unittest

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrulesstandard

Hackwar committed:
Small fixes

Hackwar committed:
Adding back in platform check

Hackwar committed:
Adding back in platform check

Hackwar committed:
Adding back in platform check

Hackwar committed:
Adding back in platform check

Hackwar committed:
Codestyle

Hackwar committed:
Codestyle

Hackwar committed:
Removing unnecessary check

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrouteradvanced

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrulesitemid

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into com_content_router_legacy

Hackwar committed:
Implementing feedback so far

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Adding "covers" notation for unittests

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrulesitemid

Hackwar committed:
Updating to latest changes to JComponentRouterView

Hackwar committed:
Merge branch '3.5-dev' of https://github.com/joomla/joomla-cms into componentrulesitemid

Conflicts:
	tests/unit/suites/libraries/cms/component/router/JComponentRouterViewTest.php

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrulesitemid

Hackwar committed:
Merge branch '3.5-dev' of https://github.com/joomla/joomla-cms into componentrulesitemid

Hackwar committed:
Fixing unittests of JComponentRouterRulesMenuTest

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into componentrulesnomenu2

Hackwar committed:
Fixing unittests of JComponentRouterRulesNomenu and adapting to JComponentRouterView

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_content_router_legacy

Conflicts:
	components/com_content/helpers/route.php

Hackwar committed:
Updating the legacy router of com_content to use JComponentRouterView

Hackwar committed:
Improving the router

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_contact_router_legacy

Conflicts:
	components/com_contact/helpers/route.php

Hackwar committed:
Modifying com_contact router to work JComponentRouterView instead of JComponentRouterAdvanced

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_newsfeed_router_legacy

Conflicts:
	components/com_newsfeeds/helpers/route.php

Hackwar committed:
Modifying com_newsfeed to work with JComponentRouterView instead of JComponentRouterAdvanced

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_users_router_legacy

Hackwar committed:
Modifying com_users to work with JComponentRouterView instead of JComponentRouterAdvanced

Hackwar committed:
Merge branch 'com_content_router_legacy' of github.com:Hackwar/joomla-cms into componentrulesstandard

Hackwar committed:
Updateing JComponentRouterRulesStandard to use JComponentRouterView

Hackwar committed:
Merge branch 'componentrulesstandard' of github.com:Hackwar/joomla-cms into componentrulesstandard

Hackwar committed:
Codestyle

Hackwar committed:
Merge branch 'componentrulesnomenu' of github.com:Hackwar/joomla-cms into componentrulesstandard

Hackwar committed:
Fixing bug in standard routing behavior

Hackwar committed:
Merge branch '3.5-dev' of https://github.com/joomla/joomla-cms into componentrulesstandard

Conflicts:
	tests/unit/suites/libraries/cms/router/JRouterSiteTest.php

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into routing

Conflicts:
	tests/unit/suites/libraries/cms/router/JRouterSiteTest.php

Hackwar committed:
Merge branch 'com_content_router_legacy' of github.com:Hackwar/joomla-cms into routing

Hackwar committed:
Merge branch 'com_contact_router_legacy' of github.com:Hackwar/joomla-cms into routing

Hackwar committed:
Merge branch 'com_newsfeed_router_legacy' of github.com:Hackwar/joomla-cms into routing

Hackwar committed:
Merge branch 'com_users_router_legacy' of github.com:Hackwar/joomla-cms into routing

Hackwar committed:
Merge branch 'componentrulesnomenu' of github.com:Hackwar/joomla-cms into routing

Hackwar committed:
Merge branch 'componentrulesstandard' of github.com:Hackwar/joomla-cms into routing

Hackwar committed:
Adding sef_advanced option

Hackwar committed:
Improving JRouterSite unittest

Hackwar committed:
Fixing Itemid bug in component menu router rule

Hackwar committed:
Merge branch '3.5-dev' of https://github.com/joomla/joomla-cms into routing

Conflicts:
	components/com_contact/helpers/route.php
	components/com_newsfeeds/helpers/route.php
	components/com_newsfeeds/router.php
	components/com_users/controllers/remind.php
	components/com_users/controllers/reset.php
	components/com_users/controllers/user.php

Hackwar committed:
Merge branch 'staging' of github.com:joomla/joomla-cms into routing

Conflicts:
	components/com_content/router.php
	components/com_users/controllers/user.php

Hackwar committed:
Fixing wrong parent ID name in component routers

Hackwar committed:
Fixing issue with categories view in router

wilsonge commented:
Can we fix the unit tests please?

Hackwar committed:
Fixing several bugs in router rules

Gitjk commented:
When I try to apply this patch with the patchtester, I always get this message:
Error
The patch could not be applied because it conflicts with a previously applied patch: administrator/components/com_config/model/form/application.xml

Tried with Hackwar's routing.zip (linked at the top of this page), latest staging, 3.5dev and master branch. Always the the same message.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8256">issues.joomla.org/joomla-cms/8256</a>.</sub>

Hackwar commented:
@Gitjk Please repair your test installation

Gitjk commented:
Just tested again with latest staging branch and my bilingual VM shop. Didn't find any issues. Looks good for me<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8256">issues.joomla.org/joomla-cms/8256</a>.</sub>

Gitjk commented:
I have tested this item :white_check_mark: successfully on a677ef33975c905bc8821bcc02e1e724fb7aa130<br /><br />Missed the 'Test this' button when I wrote my comment a few minutes ago. In my case the test with latest staging and latest VirtueMart installed (not yet released) is successful. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8256">issues.joomla.org/joomla-cms/8256</a>.</sub>

astridx commented:
That is the first time, I have a closer look on the joomla router.
I started with the instructions of issue #5546.
I change the router of com_content by replacing the described texts. 
I want to see that the right menu items are still found for the items but I only get the message 
Fatal error: Class 'JComponentRouterRulesMenu' not found in E:\Daten\git\joomla-cms\components\com_content\router.php on line 35
I searched for this class myself and I found it in JOOMLA\libraries\cms\component\router\rules\interface.php. I reverted my changes and then I went on with the instructions of #5501, because I am not sure if the instructions in #5546 are obsolete. Are this obsolete?
---
I checked my site before applying the patch and checked the URLs of com_content.
 I apply the patch.
I realized that the URLs of com_content are staying the same. I used the English sample data.
The URL of All Front End Views | Content Component was 
was
http://localhost/joomla-cms/index.php/content-component
and is after applying the patch
http://localhost/joomla-cms/index.php/content-component
---
Then I followed the instructions of #5502
The URL of All Front End Views | Contact Component
was
http://localhost/joomla-cms/index.php/contact-component
and is after applying the patch
http://localhost/joomla-cms/index.php/contact-component
---
Then I followed the instructions of #5503
The URL of All Front End Views | News Feed Component
was
http://localhost/joomla-cms/index.php/news-feeds-component
and is after applying the patch
http://localhost/joomla-cms/index.php/news-feeds-component
---
Then I followed the instructions of #5509
The URL of All Front End Views | Users Component
was
http://localhost/joomla-cms/index.php/users-component
and is after applying the patch
http://localhost/joomla-cms/index.php/users-component
I read that the PR #5509 requires #5446 to be accepted by the project before it can be applied. But now the PR #5446 is part of this PR.
---
With PR #5599 it was the same like PR #5546. I only get the error message: Fatal error: Class 'JComponentRouterRulesMenu' not found in E:\Daten\git\joomla-cms\components\com_content\router.php on line 35. Obsolete Instruction?
---
I end with the instructions of the PR 5604 where the over-righting ended in a failure message as well. Obsolete Instruction?

---

Then I went on with testing if the failure of issue #8141 is fixed with this new router.
1.	Create a category named "projects"
2.	Create a second category named p1 and make it a child of "projects"
3.	Create a third category named 1.2.3 and make it a child of p1. Save it with the default alias 1-2.3.
4.	Create a menu entry of the List All Categories type and select "projects"
5.	Go to the front end of the site and navigate to the new menu entry.
6.	Click on it so that you see subcategory p1.
7.	Click on p1 so that you see the subcategory 1.2.3.
8.	Click on 1.2.3
After activating “Use new URL routing” in the SEO Settings, this failure is fixed.

In the end I installed a small selfmade component and found no issues.

schnuti commented:
I have tested this item :white_check_mark: successfully for com_content with a new test installation.

It would though be nice to know the problems solved and known problems remaining.

I found one issue remaining concerning canonical link.

http://localhost/joomla-cms-routing/parks-home/21
canonical:
http://localhost/joomla-cms-routing/article-category-list?id=21

While this gives a correct one.
http://localhost/joomla-cms-routing/parks-home/21-x
canonical
http://localhost/joomla-cms-routing/article-category-list/21-extensions/components/21-getting-help

This probaly comes from the legacyrouter - parse and is not new!
// We check to see if an alias is given.  If not, we assume it is an article

schnuti commented:
Ooops! Forgot to set the parameter to use the new router.
What I wrote above is still valid.
With activated new router the wrong canonical links looks better ;)

http://localhost/joomla-cms-routing/parks-home/21
canonical
http://localhost/joomla-cms-routing/single-article

The page shown is the menu item Parks "Welcome! ... parks of Australia"
 
If I use this canonical link, I become the single Article of "Welcome! ... parks of Australia"

In other menu items the canonical link points to the menu item itself. Correct.

ie the default to com_content article and/or category has bin removed. Change to the better I believe.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8256">issues.joomla.org/joomla-cms/8256</a>.</sub>

schnuti commented:
BUG!
Found a bug when scrolling down the page. Below the login module.

3 times the same:
Notice: Undefined index: layout in D:\xampp\htdocs\joomla-cms-routing\libraries\cms\component\router\rules\standard.php on line 212

Check isset?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8256">issues.joomla.org/joomla-cms/8256</a>.</sub>

schnuti commented:
I'm now trying to test wih a custom component and need some hint for the setup in the router. I tried some variations of the setup for com_content.
I have a main menu (view) that is a "merge" of categories and category. ie nested categories from root to last level. The botttom level has items with links to the item view. normaly shown in the same menuItem.

com_content defines both categories and category, where category is using the categories definition.

Any help and hint?

anibalsanchez commented:
I have tested this item :white_check_mark: successfully on a677ef33975c905bc8821bcc02e1e724fb7aa130<br /><br />Applied with patchester and tested several cases, it is working OK. Is there any additional requirements for testing?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8256">issues.joomla.org/joomla-cms/8256</a>.</sub>

Hackwar committed:
Fixing notice

Hackwar committed:
Fixing unittests

Hackwar committed:
Fixing unittest for PHP7

Hackwar committed:
Codestyle

Hackwar committed:
Codestyle

Hackwar committed:
Codestyle ... again

Hackwar commented:
Hi folks, I cleaned up the code further, fixed the unittests and fixed the codestyle. Please give this another round of tests.

To test, please apply these changes and enable the new routing in the global configuration. Then check the URLs of the site. The best would be, if you could test this on a copy of a real site. The sampledata of Joomla has already been tested and works. When you test these changes, the URLs should stay the same in most cases. The only time when the URL will be changed by this is when the menu item links to a category and the content item is in a child-child-category of the linked category.

Gitjk commented:
Hi again,
Just had a short (10 Minute) look. Used latest staging Joomla and VirtueMart 3.0.12, both bilingual. After adding #8256 with the patch-tester, I didn't see any difference in urls with 'Use new URL routing' enabled/disabled. Seems to work fine at the first glance. I might have another look on Sunday.

infograf768 commented:
@test
Looks like working fine here on my multilingual test site.
Did not test the child-child situation yet.

infograf768 commented:
Please alpha order new lang strings.

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -168,6 +168,8 @@ COM_CONFIG_FIELD_PROXY_USERNAME_DESC="The username used to access the Proxy serv
 COM_CONFIG_FIELD_SECRET_DESC="This is an auto-generated, unique alphanumeric code for every Joomla installation. It is used for security functions."
 COM_CONFIG_FIELD_SECRET_LABEL="Secret"
+COM_CONFIG_FIELD_SEF_ADVANCED_LABEL="Use new URL routing"
+COM_CONFIG_FIELD_SEF_ADVANCED_DESC="This uses the new URL routing. This will change your URLs!"
```
will or may?

Hackwar replied:
You are right, it "may" change your URLs.

anibalsanchez commented:
Tested on a news site, upgraded to Joomla 3.5, installed the new router, crawled Urls with Screaming Frog and compared results. No differences found in resulting Urls.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8256">issues.joomla.org/joomla-cms/8256</a>.</sub>

wilsonge commented:
@Hackwar can you please update this PR to staging (conflicts atm). Got some more people to test on this

Hackwar committed:
Implementing component specific advanced routing setting

Hackwar committed:
Merge branch 'staging' of github.com:joomla/joomla-cms into routing

Conflicts:
	components/com_content/router.php

Hackwar committed:
Reverting erronous commit

Hackwar commented:
I've updated the branch. Please notice that I've moved the setting for advanced routing from the global configuration to the component configuration for each specific component.

wilsonge commented:
Merged into 3.6.x branch :)

810 commented:
UsersHelperRoute::getLoginRoute(); is set to deprecated, what's the new call, or are the deprecated tags wrong on the j3.6/7 branch, because on the 4.0 branch its still there

wilsonge commented:
4.0 branch is currently based on staging not the 3.7 branch :)
